application-ideas-api/src/main/java/com/xwiki/ideas/IdeasManager.java

```diff
@@ -0,0 +1,49 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.ideas;
+
+import org.xwiki.component.annotation.Role;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.stability.Unstable;
+
+/**
+ * @version $Id$
+ * @since 1.4
+ */
+@Role
+@Unstable
+public interface IdeasManager
+{
+    /**
+     * @param documentReference the document that we want to get the URL for
+     * @return the Ideas REST URL specific to the document
+     */
+    String getRestUrl(DocumentReference documentReference);
+
+    /**
+     *
+     * @param documentReference a reference to a document that contains an Idea
+     * @param voteType the type of the vote that the user wants to cast
```

Comment:
OK, I checked the implementation of ``vote()`` and this is not a good / clean API. It's important to make the distinction between how you implement the voting at the UI level, using a form that gets submitted to some URL that has some parameters specifying the type of action, etc. and the Java API. The Java API needs to be **clean** and **typed**. I don't think there are endless actions that you can do on an idea. On the contrary, there are a few, well known actions that have to be exposed in the API. So you can either have:

* ``voteFor(DocumentReference ideaReference)``
* ``voteAgainst(DocumentReference ideaReference)``
* ``removeVote(DocumentReference ideaReference)``

or

* ``vote(DocumentReference ideaReference, boolean for)``
* ``removeVote(DocumentReference ideaReference)``

or

* ``vote(DocumentReference ideaReference, Boolean for)``

(passing ``null`` to remove the vote). Depends how explicit you want to be. In any case, the generic string action makes sense when you want it to be the role hint of a component (``IdeaAction``) because you want to support extensible actions, which is too far fetched for our needs.